api/views/canteen.py

```diff
@@ -0,0 +1,262 @@
+
+
+class AddManagerView(APIView):
+    permission_classes = [permissions.IsAuthenticated]
+
+    def post(self, request, *args, **kwargs):
+        try:
+            email = request.data.get("email")
+            validate_email(email)
+            canteen_id = request.data.get("canteen_id")
+            canteen = request.user.canteens.get(id=canteen_id)
+            try:
+                user = get_user_model().objects.get(email=email)
+                canteen.managers.add(user)
+            except get_user_model().DoesNotExist:
+                with transaction.atomic():
+                    pm = ManagerInvitation(canteen_id=canteen.id, email=email)
+                    pm.save()
+                AddManagerView._send_invitation_email(pm)
+            return _respond_with_team(canteen)
+        except ValidationError as e:
+            logger.error(f"Attempt to add manager with invalid email {email}")
+            logger.exception(e)
+            return JsonResponse(
+                {"error": "Invalid email"}, status=status.HTTP_400_BAD_REQUEST
+            )
+        except Canteen.DoesNotExist as e:
+            logger.error(f"Attempt to add manager to unexistent canteen {canteen_id}")
+            logger.exception(e)
+            return JsonResponse(
+                {"error": "Invalid canteen id"}, status=status.HTTP_404_NOT_FOUND
+            )
+        except IntegrityError as e:
+            logger.error(
+                f"Attempt to add existing manager with email {email} to canteen {canteen_id}"
+            )
+            logger.exception(e)
+            return _respond_with_team(canteen)
+        except Exception as e:
+            logger.error("Exception ocurred while inviting a manager to canteen")
+            logger.exception(e)
+            return JsonResponse(
+                {"error": "An error has ocurred"},
+                status=status.HTTP_500_INTERNAL_SERVER_ERROR,
+            )
+
+    @staticmethod
+    def _send_invitation_email(manager_invitation):
+        try:
+            template = "auth/manager-invitation"
+            context = {
+                "canteen": manager_invitation.canteen.name,
+                "protocol": "https" if settings.SECURE_SSL_REDIRECT else "http",
+                "domain": settings.HOSTNAME,
+            }
+            send_mail(
+                subject="Invitation à gérer une cantine sur ma cantine",
+                message=render_to_string(f"{template}.txt", context),
+                from_email=settings.DEFAULT_FROM_EMAIL,
+                html_message=render_to_string(f"{template}.html", context),
+                recipient_list=[manager_invitation.email],
+                fail_silently=False,
+            )
+        except Exception as e:
+            logger.error("The manager invitation email could not be sent:")
+            logger.exception(e)
+            raise Exception("Error occurred : the mail could not be sent.") from e
```

Comment:
En ajoutant `from e` on utilise l'_exception chaining_ de Python (https://docs.python.org/3/reference/simple_stmts.html#the-raise-statement)